### PR TITLE
Update docs on iree-opt to use a file and options that still exist.

### PR DESCRIPTION
**Pull request by ScottTodd**
Open to suggestions for a simpler file / set of passes that we expect to be more stable.

### CODE REVIEW

**Comment by benvanik** [comment]
+1 to `-iree-transformation-pipeline` & co - anything more fine-grained is (effectively) an implementation detail for compiler hackers of specific backends that is better documented there. Other ones that may be useful would be the one level down from iree-transformation-pipeline:
https://github.com/google/iree/blob/0ea7b14cd3c44b41a30b5f764c7c572043991607/iree/compiler/Translation/IREEVM.cpp#L68-L71

(`-iree-flow-transformation-pipeline`, hal, vm, etc)

**Comment by ScottTodd** [comment]
> +1 to `-iree-transformation-pipeline` & co - anything more fine-grained is (effectively) an implementation detail for compiler hackers of specific backends that is better documented there. Other ones that may be useful would be the one level down from iree-transformation-pipeline:
> https://github.com/google/iree/blob/0ea7b14cd3c44b41a30b5f764c7c572043991607/iree/compiler/Translation/IREEVM.cpp#L68-L71
> 
> (`-iree-flow-transformation-pipeline`, hal, vm, etc)

I had been thinking with these docs that `iree-translate` would be more user-oriented and `iree-opt` would be more developer / compiler-hacker oriented. I wanted the `iree-opt` example to be focused on a single (tiny) pass where you could easily see the diff between the input and output and associate that with the name of the pass ("perform inlining", "fuse these two functions", "add attributes", those types of passes).

The larger pass pipelines also work in docs, but they show something different. There's more going on, and they could be useful to people plugging systems together, but they don't teach the system fundamentals as directly.

**Comment by GMNGeoffrey** [comment]
> I had been thinking with these docs that `iree-translate` would be more user-oriented and `iree-opt` would be more developer / compiler-hacker oriented. I wanted the `iree-opt` example to be focused on a single (tiny) pass where you could easily see the diff between the input and output and associate that with the name of the pass ("perform inlining", "fuse these two functions", "add attributes", those types of passes).
> 
> The larger pass pipelines also work in docs, but they show something different. There's more going on, and they could be useful to people plugging systems together, but they don't teach the system fundamentals as directly.

I agree this should be the common case for opt and would be more useful to document. My vote would be to pick an existing test case for a relatively stable transform and then IfChange/ThenChange to note that it's referenced in the documentation and the documentation should be updated if it is.

**Comment by ScottTodd** [comment]
> > I had been thinking with these docs that `iree-translate` would be more user-oriented and `iree-opt` would be more developer / compiler-hacker oriented. I wanted the `iree-opt` example to be focused on a single (tiny) pass where you could easily see the diff between the input and output and associate that with the name of the pass ("perform inlining", "fuse these two functions", "add attributes", those types of passes).
> > The larger pass pipelines also work in docs, but they show something different. There's more going on, and they could be useful to people plugging systems together, but they don't teach the system fundamentals as directly.
> 
> I agree this should be the common case for opt and would be more useful to document. My vote would be to pick an existing test case for a relatively stable transform and then IfChange/ThenChange to note that it's referenced in the documentation and the documentation should be updated if it is.

How about `-iree-vm-mark-public-symbols-exported` on https://github.com/google/iree/blob/master/iree/compiler/Dialect/VM/Transforms/test/mark_public_symbols_exported.mlir? That's one of the smallest / simplest that I could find. We could even document both the small passes like that and the larger collections of passes like `iree-transformation-pipeline`.

```
λ .\iree\tools\iree-opt -iree-vm-mark-public-symbols-exported %CD%/../iree/iree/compiler/Dialect/VM/Transforms/test/mark_public_symbols_exported.mlir -print-ir-before-all
// *** IR Dump Before mlir::iree_compiler::IREE::VM::MarkPublicSymbolsExportedPass ***


module {
  func @private_symbol() attributes {sym_visibility = "private"}
  func @public_symbol()
}



module {
  func @private_symbol() attributes {sym_visibility = "private"}
  func @public_symbol() attributes {iree.module.export}
}
```

**Comment by ScottTodd** [comment]
Made another attempt, incorporating some ideas from the discussion here. PTAL

**Comment by GMNGeoffrey** [comment]
I think `iree.module.export` is something slated for deletion since MLIR has real symbol visibility now, so probably not the best example :-D

What about [drop compiler hints](https://sourcegraph.com/github.com/google/iree/-/blob/iree/compiler/Dialect/IREE/Transforms/test/drop_compiler_hints.mlir)? Simple pass that just deletes a few ops. No current plans to change it. I'm not caught up enough on the other compiler projects to know what else is stable except that I would avoid anything with dynamic shapes, stream formation, or dispatchability :-D

File: iree/compiler/Dialect/IREE/Transforms/test/drop_compiler_hints.mlir [hunk 1]
```diff
@@ -1,5 +1,8 @@
 // RUN: iree-opt -split-input-file -iree-drop-compiler-hints %s | IreeFileCheck --implicit-check-not="iree.do_not_optimize" %s
 
+// This file is used as an example in docs/developer_overview.md.
+// If you move or delete it, please update the documentation accordingly.
+
```
Comment (GMNGeoffrey) [inline iree/compiler/Dialect/IREE/Transforms/test/drop_compiler_hints.mlir:5]: I would still prefer IfChange/ThenChange around the `RUN` line, but ok

File: docs/developer_overview.md [hunk 2]
```diff
@@ -68,24 +68,34 @@ translations, and other transformations step by step.
 [mlir-opt](https://github.com/llvm/llvm-project/tree/master/mlir/tools/mlir-opt)
 and runs sets of IREE's compiler passes on `.mlir` input files. See "conversion"
 in [MLIR's Glossary](https://mlir.llvm.org/getting_started/Glossary/#conversion)
-for more information.
+for more information. Compiler "passes" can range from isolated manipulations to
+broad pipelines that encompass a sequence of steps.
```
Comment (GMNGeoffrey) [inline docs/developer_overview.md:72]: A pipeline isn't a pass. It's a series of passes

```suggestion
for more information. Transformations performed by `iree-opt` can range from
individual passes performing isolated manipulations to broad pipelines that
encompass a sequence of steps.
```

**Pull request merged**